### ISSUE TITLE
gzip decoding not working/gzip headers not supported

### ISSUE THREAD

**Issue by Hitokun** [celery/kombu#707]
Im posting gzip encoded messages and I want to take advantage of the apparent built in decoding capabilities of kombu, but it is not letting me use gzip. zlib is recognized by kombu, but apparently it tries to inflate instead of using gzip decompression (the error is `kombu.exceptions.DecodeError: Error -3 while decompressing data: incorrect header check`)

kombu==4.0.2

```
/* index.js */
exchangeOptions = {
    'contentType': 'application/x-gzip',
    'contentEncoding': 'gzip',
    'appId': 'statsdAMQP',
    'deliveryMode': 2
};
...
...
var post_stats = function (payload, callback) {
    try {
        for (var key in payload) {
            var data = payload[key];
            channel.publish(options.exchange, data.metric, zlib.gzipSync(data.result), exchangeOptions);
            if (debug) {
                util.log("Published: " + data.result);
            }
        }
        amqpStats.last_flush = Math.round(new Date().getTime() / 1000);
        callback();
    } catch (e) {
        if (debug) {
            util.log(e);
        }
        amqpStats.last_exception = Math.round(new Date().getTime() / 1000);
    }
};
```

```
// dummy_consumer.py
from kombu import Connection

with Connection('amqp://guest:guest@172.17.0.3:5672//') as conn:
    simple_queue = conn.SimpleQueue('statsd')
    message = simple_queue.get(block=True, timeout=1)
    print ("Received %s" % message.payload)
    message.ack()
    simple_queue.close()
```

```
$ python dummy_consumer.py 
Traceback (most recent call last):
  File "dummy_consumer.py", line 5, in <module>
    message = simple_queue.get(block=True, timeout=1)
  File "/usr/lib/python2.7/site-packages/kombu/simple.py", line 47, in get
    timeout=timeout and remaining)
  File "/usr/lib/python2.7/site-packages/kombu/connection.py", line 301, in drain_events
    return self.transport.drain_events(self.connection, **kwargs)
  File "/usr/lib/python2.7/site-packages/kombu/transport/pyamqp.py", line 103, in drain_events
    return connection.drain_events(**kwargs)
  File "/usr/lib/python2.7/site-packages/amqp/connection.py", line 464, in drain_events
    return self.blocking_read(timeout)
  File "/usr/lib/python2.7/site-packages/amqp/connection.py", line 469, in blocking_read
    return self.on_inbound_frame(frame)
  File "/usr/lib/python2.7/site-packages/amqp/method_framing.py", line 88, in on_frame
    callback(channel, msg.frame_method, msg.frame_args, msg)
  File "/usr/lib/python2.7/site-packages/amqp/connection.py", line 473, in on_inbound_method
    method_sig, payload, content,
  File "/usr/lib/python2.7/site-packages/amqp/abstract_channel.py", line 142, in dispatch_method
    listener(*args)
  File "/usr/lib/python2.7/site-packages/amqp/channel.py", line 1613, in _on_basic_deliver
    fun(msg)
  File "/usr/lib/python2.7/site-packages/kombu/messaging.py", line 617, in _receive_callback
    decoded = None if on_m else message.decode()
  File "/usr/lib/python2.7/site-packages/kombu/message.py", line 192, in decode
    self._decoded_cache = self._decode()
  File "/usr/lib/python2.7/site-packages/kombu/message.py", line 197, in _decode
    self.content_encoding, accept=self.accept)
  File "/usr/lib/python2.7/site-packages/kombu/serialization.py", line 267, in loads
    return _decode(data, content_encoding)
  File "/usr/lib64/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/lib/python2.7/site-packages/kombu/serialization.py", line 54, in _reraise_errors
    reraise(wrapper, wrapper(exc), sys.exc_info()[2])
  File "/usr/lib/python2.7/site-packages/kombu/serialization.py", line 50, in _reraise_errors
    yield
  File "/usr/lib/python2.7/site-packages/kombu/serialization.py", line 267, in loads
    return _decode(data, content_encoding)
kombu.exceptions.DecodeError: unknown encoding: gzip
```

A workaround is to set the content_encoding to binary and then `decoded_payload = zlib.decompress(message.payload, 16+zlib.MAX_WBITS)`

**Comment by Hitokun** [comment]
So, I discovered that setting `compression` in the message headers with either gzip or zlib makes Kombu to do it's thing consistently. I was still getting the incorrect header check, so I changed gzip compression for deflate and it did the trick. This is at least misleading. checking the source code, I see decompressor using a standard zlib.decompress. Im running stock Python 2.7.13 in Fedora 25, and in my workaround I need to set  the _wbits _ argument so it can actually read the gzip header. I don't know if there are zlib versions that can handle this automatically, but for now it seems gzip support is only partially true (in the sense that the compression algorithm is the same, but the gzip headers are not really supported).
